File: src/cpp/src/gravity.cc
#include "mujoco_orbit/gravity.h"

#include "mujoco_orbit/math_utils.h"

namespace mujoco_orbit {

void point_mass_accel(const double r_eci[3], double out_a[3], double gm) {
  const double r = detail::norm3(r_eci);
  if (r == 0.0) {
    detail::zero3(out_a);
    return;
  }
  const double factor = -gm / (r * r * r);
  detail::scale3(r_eci, factor, out_a);
}

void j2_accel(const double r_eci[3], double out_a[3], double gm, double j2, double r_eq) {
  const double x = r_eci[0];
  const double y = r_eci[1];
  const double z = r_eci[2];
  const double r = detail::norm3(r_eci);
  if (r == 0.0) {
    detail::zero3(out_a);
    return;
  }
  const double factor = 1.5 * j2 * gm * r_eq * r_eq / std::pow(r, 5);
  const double z_r2 = (z / r) * (z / r);
  out_a[0] = factor * x * (5.0 * z_r2 - 1.0);
  out_a[1] = factor * y * (5.0 * z_r2 - 1.0);
  out_a[2] = factor * z * (5.0 * z_r2 - 3.0);
}

void total_accel(const double r_eci[3], double out_a[3], bool use_j2, double gm, double j2, double r_eq) {
  point_mass_accel(r_eci, out_a, gm);
  if (!use_j2) {
    return;
  }
  double a_j2[3];
  j2_accel(r_eci, a_j2, gm, j2, r_eq);
  detail::add3(out_a, a_j2, out_a);
}

void total_accel(
    const double r_eci[3],
    double out_a[3],
    bool use_j2,
    const CentralBodySpecNative& central_body) {
  total_accel(
      r_eci,
      out_a,
      use_j2,
      central_body.gm,
      central_body.j2,
      central_body.radius);
}

void encke_point_mass_relative_accel(
    const double rho[3],
    const double r_chief[3],
    double out_a[3],
    double gm) {
  const double rc2 = detail::dot3(r_chief, r_chief);
  if (rc2 == 0.0) {
    detail::zero3(out_a);
    return;
  }
  const double rc = std::sqrt(rc2);

  // sigma = rho . (2 r_chief + rho) / r_chief^2, computed without forming
  // 2 r_chief + rho explicitly so we can keep it stable for tiny rho.
  const double sigma =
      (2.0 * detail::dot3(rho, r_chief) + detail::dot3(rho, rho)) / rc2;

  // f(sigma) = 1 - (1+sigma)^{-3/2}, evaluated as
  // sigma * (3 + 3 sigma + sigma^2) / [(1 + (1+sigma)^{3/2}) * (1+sigma)^{3/2}]
  // to avoid cancellation when sigma -> 0.
  const double one_plus_sigma = 1.0 + sigma;
  const double one_plus_sigma_3_2 = one_plus_sigma * std::sqrt(one_plus_sigma);
  const double f =
      sigma * (3.0 + 3.0 * sigma + sigma * sigma) /
      ((1.0 + one_plus_sigma_3_2) * one_plus_sigma_3_2);

  const double scale = -gm / (rc2 * rc);
  for (int i = 0; i < 3; ++i) {
    out_a[i] = scale * (rho[i] - f * (r_chief[i] + rho[i]));
  }
}

void relative_accel(
    const double rho[3],
    const double r_chief[3],
    double out_a[3],
    bool use_j2,
    const CentralBodySpecNative& central_body) {
  encke_point_mass_relative_accel(rho, r_chief, out_a, central_body.gm);
  if (!use_j2) {
    return;
  }
  // J2 differential is much smaller than the point-mass term, so direct
  // subtraction is numerically fine.
  double r_body[3];
  detail::add3(r_chief, rho, r_body);
  double j2_body[3];
  double j2_chief[3];
  j2_accel(r_body, j2_body, central_body.gm, central_body.j2, central_body.radius);
  j2_accel(r_chief, j2_chief, central_body.gm, central_body.j2, central_body.radius);
  for (int i = 0; i < 3; ++i) {
    out_a[i] += j2_body[i] - j2_chief[i];
  }
}

}  // namespace mujoco_orbit

File: tests/mujoco_orbit/reference/orbit/gravity.py
"""Gravitational acceleration models.

Units throughout: km, s.
Ported from OrbitX dynamics/gravity.py (Vallado formulation).
"""

from __future__ import annotations

import numpy as np

from mujoco_orbit.constants import GM_EARTH, J2_EARTH, R_EARTH


def point_mass_accel(r_eci: np.ndarray) -> np.ndarray:
    """Two-body gravitational acceleration.

    Args:
        r_eci: position in ECI, shape (3,) km

    Returns:
        acceleration, shape (3,) km/s^2
    """
    r = np.linalg.norm(r_eci)
    return -GM_EARTH / r**3 * r_eci


def j2_accel(r_eci: np.ndarray) -> np.ndarray:
    """J2 perturbation acceleration in ECI.

    Args:
        r_eci: position in ECI, shape (3,) km

    Returns:
        J2 acceleration, shape (3,) km/s^2
    """
    x, y, z = r_eci
    r = np.linalg.norm(r_eci)
    factor = (3.0 / 2.0) * J2_EARTH * GM_EARTH * R_EARTH**2 / r**5
    z_r2 = (z / r) ** 2
    ax = factor * x * (5.0 * z_r2 - 1.0)
    ay = factor * y * (5.0 * z_r2 - 1.0)
    az = factor * z * (5.0 * z_r2 - 3.0)
    return np.array([ax, ay, az])


def total_accel(r_eci: np.ndarray, use_j2: bool = True) -> np.ndarray:
    """Sum of point-mass and optional J2 acceleration."""
    a = point_mass_accel(r_eci)
    if use_j2:
        a = a + j2_accel(r_eci)
    return a


def encke_point_mass_relative_accel(rho: np.ndarray, r_chief: np.ndarray) -> np.ndarray:
    """Encke's identity for the two-body differential ``g_pm(r_chief+rho) - g_pm(r_chief)``.

    Avoids the catastrophic subtraction of two near-equal large gravity
    vectors when ``||rho|| << ||r_chief||`` (paper eq. 'encke').

    Args:
        rho: chief-relative position, shape (3,) km.
        r_chief: chief absolute ECI position, shape (3,) km.

    Returns:
        Differential acceleration, shape (3,) km/s^2.
    """
    rho = np.asarray(rho, dtype=float)
    r_chief = np.asarray(r_chief, dtype=float)
    rc2 = float(r_chief @ r_chief)
    if rc2 == 0.0:
        return np.zeros(3)
    rc = np.sqrt(rc2)

    sigma = (2.0 * float(rho @ r_chief) + float(rho @ rho)) / rc2
    one_plus_sigma = 1.0 + sigma
    one_plus_sigma_3_2 = one_plus_sigma * np.sqrt(one_plus_sigma)
    f_sigma = (
        sigma
        * (3.0 + 3.0 * sigma + sigma * sigma)
        / ((1.0 + one_plus_sigma_3_2) * one_plus_sigma_3_2)
    )

    return -GM_EARTH / (rc2 * rc) * (rho - f_sigma * (r_chief + rho))


def relative_accel(
    rho: np.ndarray,
    r_chief: np.ndarray,
    use_j2: bool = True,
) -> np.ndarray:
    """Differential gravity ``g(r_chief+rho) - g(r_chief)``.

    Uses Encke's identity for the two-body term (well-conditioned at single
    precision) and direct subtraction for the much smaller J2 perturbation.
    """
    a = encke_point_mass_relative_accel(rho, r_chief)
    if use_j2:
        r_chief = np.asarray(r_chief, dtype=float)
        rho = np.asarray(rho, dtype=float)
        a = a + j2_accel(r_chief + rho) - j2_accel(r_chief)
    return a
